File: ecosystems/swift-package/Sources/AiTextCleanerFormattingResidue/FormattingResidue.swift
import Foundation

public enum FormattingResidue {
    public static func clean(_ text: String) -> String {
        let blockPattern = #"</?(?:address|article|blockquote|br|div|li|p|pre|section)\b[^>]*>"#
        let tagPattern = #"<[^>]*>"#
        var cleaned = text.replacingOccurrences(of: blockPattern, with: "\n", options: .regularExpression)
            .replacingOccurrences(of: tagPattern, with: "", options: .regularExpression)
        cleaned = decodeEntities(cleaned)
            .replacingOccurrences(of: "“", with: "\"")
            .replacingOccurrences(of: "”", with: "\"")
            .replacingOccurrences(of: "‘", with: "'")
            .replacingOccurrences(of: "’", with: "'")
            .replacingOccurrences(of: "–", with: "-")
            .replacingOccurrences(of: "—", with: "-")
            .replacingOccurrences(of: "…", with: "...")
            .replacingOccurrences(of: #"[ \t]+\n"#, with: "\n", options: .regularExpression)
            .replacingOccurrences(of: #"\n{3,}"#, with: "\n\n", options: .regularExpression)
        return cleaned.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func decodeEntities(_ text: String) -> String {
        text.replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#39;", with: "'")
            .replacingOccurrences(of: "&apos;", with: "'")
            .replacingOccurrences(of: "&nbsp;", with: " ")
    }
}

File: php/src/AiTextCleaner/FormattingResidue.php
<?php

declare(strict_types=1);

namespace AiTextCleaner;

final class FormattingResidue
{
    private function __construct() {}

    public static function clean(string $text): string
    {
        $text = preg_replace('/<\s*(address|article|blockquote|br|div|li|p|pre|section)\b[^>]*>/i', "\n", $text) ?? $text;
        $text = preg_replace('/<\s*\/(address|article|blockquote|br|div|li|p|pre|section)\s*>/i', "\n", $text) ?? $text;
        $text = strip_tags($text);
        $text = html_entity_decode($text, ENT_QUOTES | ENT_HTML5, 'UTF-8');
        $text = strtr($text, ['“' => '"', '”' => '"', '‘' => "'", '’' => "'", '–' => '-', '—' => '-', '…' => '...']);
        $text = preg_replace('/[ \t]+\n/u', "\n", $text) ?? $text;
        $text = preg_replace('/\n{3,}/u', "\n\n", $text) ?? $text;
        return trim($text);
    }
}
